### ISSUE TITLE
When I deploy to Vercel / Next 13 and I change routes to a page I get "DOMException: Failed to execute 'showModal' on 'HTMLDialogElement': The element is not in a Document."

### ISSUE THREAD

**Issue by kizeesmack** [rpearce/react-medium-image-zoom#429]
When the page is refreshed the modal works. When the page is visited through a link it does not work and the error comes up.

"DOMException: Failed to execute 'showModal' on 'HTMLDialogElement': The element is not in a Document."

Steps to reproduce / narrow test
1. Tried both zoom and controlled zoom, both are having issues.
2. Confirm that the modal does not work and the error comes only on route change , the initial route has no zoom instance.
3. Refresh on route with zoom and it works fine. 

using React latest 18+ and Next js 13 app directory

**Comment by kizeesmack** [comment]
just in addition on initial home page load with a component that has <Zoom> in it 
<div data-rmiz-portal> is not in the html tree. When it is refreshed and visited through another route it appears. 

This is does not happen in development mode only production / npm start.

**Comment by rpearce** [comment]
Hi, @kizeesmack, and I'm sorry you're having issues.

Are you saying that when the page is loaded, you are immediately trying to zoom the image (which opens the modal dialog), but the error occurs?

It would be immensely helpful if you could share a functioning example that reproduces the issue. That saves me from trying to guess exactly what you're doing, and it will let us figure this out faster.

**Comment by steven-tey** [comment]
Running into this issue as well @rpearce, but it looks like it happens intermittently. Will try to see if I can figure out why it's happening

**Comment by steven-tey** [comment]
Here are a few examples for reference:

- Doesn't work: https://dub.sh/help/article/what-is-dub
- Works: https://dub.sh/help/article/pro-plan

https://github.com/rpearce/react-medium-image-zoom/assets/28986134/83bcb88c-1d18-4f15-a0e9-096b53254d2a

**Comment by rpearce** [comment]
@steven-tey Do you have the time to create an example project that demonstrates the issue? My spouse's health issues are extremely bad today, but if I've got someplace to start, I can start looking at it tonight.

**Comment by steven-tey** [comment]
Made a minimal repro here: https://rmiz-bug.vercel.app/

Code: https://github.com/uncurated-tests/rmiz-bug

It seems to work correctly though so I'm going to keep investigating to see what's causing this issue!

**Comment by rpearce** [comment]
Yes, something tiny that demonstrates the issue will probably mean a quick fix!

**Comment by steven-tey** [comment]
Ahhh I figured it out! This happens when you use `react-medium-image-zoom` with the Edge Runtime – getting this in the logs:

```
EvalError: Code generation from strings disallowed for this context
    at (node_modules/next-contentlayer/dist/hooks/useMDXComponent.js:9:15)
    at (node_modules/next-contentlayer/dist/hooks/useMDXComponent.js:13:31)
```

Made a minimal repro here: https://rmiz-bug.vercel.app/

https://github.com/rpearce/react-medium-image-zoom/assets/28986134/ffbbdb1e-ee60-449c-8d27-10b0f3bbb6c5

Here's the code: https://github.com/uncurated-tests/rmiz-bug

Changing the runtime to `nodejs` worked as a temporary fix for now!

**Comment by rpearce** [comment]
Ok, that's a great investigation and breakdown. Please let me know if something needs adjusting over on this side.

**Comment by flosrn** [comment]
Hi @steven-tey , have you found a solution? I am facing the same issue

**Comment by ronald-das** [comment]
Can confirm, facing the same issue with nextjs `13.4.19` and react-medium-image-zoom: `5.1.8` , @steven-tey saw your tweet https://twitter.com/steventey/status/1684607625240150016 what are you guys using? couldn't find this library in https://github.com/vercel/next.js/blob/canary/package.json

**Comment by tszhong0411** [comment]
@ronald-das The documentation for Next.js is not open source.

**Comment by thiskevinwang** [comment]
@kizeesmack are you possibly using this via MDX?

If so, try adding https://github.com/remarkjs/remark-unwrap-images as a `remarkPlugin`.

I was getting the same error locally... and I _think_ it was predicated by a hydration mismatch... and I _think_ this was due to MDX wrapping images in a `<p>` tag. (see https://github.com/mdx-js/mdx/issues/208 for details)

```mdx
Hello

![](/foo.webp)

That image was getting wrapped in a P tag and I was seeing hydration mismatches
```

I installed that remark plugin, hydration errors went away, and I didn't get the Zoom error anymore. 🎉

As an aside, you can also see on https://nextjs.org/docs/app/building-your-application/rendering that the images (`<figure>`) are not nested with a `<p>` tag.

**Comment by ilhamsa1** [comment]
I am facing the same issue. Have you found a solution, or do you have any other library recommendations?

**Comment by rpearce** [comment]
Y'all feel free to submit a PR if you think you know how to solve it on this side (as opposed to a possible bug in Next.js, as mentioned above).

**Comment by rpearce** [comment]
@steven-tey 

> This looks more of a Next.js bug so I've surfaced this to the Next.js team – we're looking into it!

Have there been any updates on the Vercel side about this?

**Comment by sgup** [comment]
Seeing the same error, 

"next": "14.0.4",
"react-medium-image-zoom": "5.1.9",

**Comment by rpearce** [comment]
If someone can provide an example repo of this breaking, I'll re-investigate to see what can be done on this side. 🙏

**Comment by u3u** [comment]
When I use Astro with View Transitions, I encounter the same problem. After looking at the source code, I found that `elDialogContainer` is initialized outside of the component's lifecycle. This means it will only be created once when `react-medium-image-zoom` is first imported. So when the page DOM changes, the `elDialogContainer` element is lost, so the `dialog` element will not be created again for the second time.

https://github.com/rpearce/react-medium-image-zoom/blob/bf6af767166df5b2eb9f0cf605a3b515006c7777/source/Controlled.tsx#L31-L37

https://github.com/rpearce/react-medium-image-zoom/blob/bf6af767166df5b2eb9f0cf605a3b515006c7777/source/Controlled.tsx#L304-L322

**Comment by u3u** [comment]
https://github.com/rpearce/react-medium-image-zoom/assets/20062482/eb7eb17b-8e62-46bd-ab9d-422fc5b5bebb

@rpearce Yes, change pages in Astro will rebuild the content within all `<html>` elements. This is a reproduction: https://qwq.cat/tools/file-to-base64

In the Next.js framework, although the content in `<body>` will not be deleted, directly refreshing the page reveals that there is no `elDialogContainer` element created within `<body>`. I believe this is the same issue because the execution timing of creating containers is during module import (as they are written at the top level of modules rather than component lifecycles), and in server-side rendering, they are not created at all due to the condition `typeof document !== 'undefined'`. I think changing it to runtime import can solve the problem in Next.js.

```ts
const Zoom = dyanmic(() => import('react-medium-image-zoom'), { ssr: false })
```

I think #462 can fix it.

**Comment by rpearce** [comment]
@u3u Thank you for the recording and the explanation. This all makes sense to me, and we'll get this issue resolved today or tomorrow!

**Comment by rpearce** [comment]
This issue should now be resolved in `react-medium-image-zoom@5.1.10`

@kizeesmack @thiskevinwang @u3u @tszhong0411 — All of you have been added as project contributors in the README thanks to your bug identification, helpfulness, and code delivery. _Thank you_ so much for your help.
